services/queue/src/api.js

```diff
@@ -300,6 +300,132 @@ builder.declare({
   return res.reply(result);
 });
 
+/** Cancel all tasks in a group */
+builder.declare({
+  method: 'post',
+  route: '/task-group/:taskGroupId/cancel',
+  name: 'cancelTaskGroup',
+  scopes: {
+    AnyOf: [
+      'queue:cancel-task-group:<schedulerId>/<taskGroupId>',
+      'queue:cancel-task-in-project:<projectId>',
+    ],
+  },
+  stability: APIBuilder.stability.experimental,
+  category: 'Tasks',
+  input: undefined,
+  output: 'cancel-task-group-response.yml',
+  title: 'Cancel Task Group',
+  description: [
+    'This method will cancel all tasks with a given `taskGroupId` that are not resolved yet.',
+    'That means all tasks in either `unscheduled`, `pending` or `running` states.',
+    'Behaviour is similar to the `cancelTask` method.',
+    '',
+    '**Remark** this operation does not guarantee that new tasks created for this `taskGroupId`',
+    'would automatically be rejected. For example when running task keeps creating new tasks.',
+  ].join('\n'),
+}, async function(req, res) {
+  const taskGroupId = req.params.taskGroupId;
+  // allow query.reasonResolved ? cancel | superseded
+
+  const [taskGroup] = await this.db.fns.get_task_group(taskGroupId);
+  if (!taskGroup) {
+    return res.reportError('ResourceNotFound',
+      'No task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      },
+    );
+  }
+
+  // in order to authorize call we would need to fetch a task to retrieve schedulerId and projectId
+  const someTaskInGroup = Task.fromDbRows(
+    await this.db.fns.get_tasks_by_task_group_projid(taskGroupId, 1, 0),
+  );
+  if (!someTaskInGroup) {
+    return res.reportError('ResourceNotFound',
+      'No tasks found for task-group with taskGroupId: `{{taskGroupId}}`', {
+        taskGroupId,
+      },
+    );
+  }
+
+  await req.authorize({
+    taskGroupId,
+    schedulerId: someTaskInGroup.schedulerId,
+    projectId: someTaskInGroup.projectId,
+  });
+
+  // naïve approach
+  const allTasks = await this.db.fns.get_tasks_by_task_group_projid(taskGroupId, null, null);
```

Comment:
This is a simple approach to fetch all tasks for the group and then cancel those that are not yet resolved and have deadline in the future. 

Each task that was cancelled emits corresponding messages to the queue

Reply:
Although more efficient to operate in the database directly rather than to call queue.cancelTask multiple times, I'm worried operating on the db directly could introduce the risk that the scope handling of queue.cancelTask and queue.cancelTaskGroup could deviate over time. So this concern isn't so much with the current risk, more about future proofing that the two mechanisms don't deviate in terms of scope requirements.

I'm curious how much performance difference there is between operating directly on the database, and calling queue.cancelTask API instead. Perhaps the overhead isn't so great, and could ensure that the two methods don't diverge?